File: frontend/src/types/smsAccount.ts
/** Mirrors backend/app/schemas/sms_account.py. Admin-only. */

export type SmsProvider = 'kwtsms' | 'unifonic' | 'smsala' | 'custom'

export interface SmsAccount {
  id: number | null
  provider: SmsProvider
  sender_id: string
  api_url: string
  api_username: string
  /** Never the real secret itself -- just whether one is set. */
  has_secret: boolean
  test_mode: boolean
  is_active: boolean
  last_tested_at: string | null
  last_test_ok: boolean | null
  last_test_error: string | null
}

export interface SmsAccountFormValues {
  provider: SmsProvider
  sender_id: string
  api_url: string
  api_username: string
  /** Empty string = leave the saved secret untouched unless the user
   * explicitly opted to change it -- see SmsTab. */
  api_secret?: string
  test_mode: boolean
  is_active: boolean
}

export interface SmsProviderPreset {
  label: string
  api_url: string
  username_label: string
  secret_label: string
  note: string
}

export type SmsProviderPresets = Record<string, SmsProviderPreset>

export interface SmsTestResult {
  ok: boolean
  message: string
}

File: backend/app/schemas/sms_account.py
from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class SmsAccountOut(BaseModel):
    id: int | None = None
    provider: str
    sender_id: str
    api_url: str
    api_username: str
    has_secret: bool  # never the real secret itself
    test_mode: bool
    is_active: bool
    last_tested_at: datetime | None = None
    last_test_ok: bool | None = None
    last_test_error: str | None = None


class SmsAccountUpdate(BaseModel):
    provider: str = Field(default="kwtsms", pattern="^(kwtsms|unifonic|smsala|custom)$")
    sender_id: str = Field(default="", max_length=20)
    api_url: str = Field(default="", max_length=255)
    api_username: str = Field(default="", max_length=255)
    # Omitted/None keeps the previously saved secret (same convention as
    # email_account's password field). Empty string explicitly clears it.
    api_secret: str | None = None
    test_mode: bool = True
    is_active: bool = True

    @field_validator("*", mode="before")
    @classmethod
    def _strip_strings(cls, v):
        return v.strip() if isinstance(v, str) else v


class SmsTestRequest(BaseModel):
    phone_number: str = Field(min_length=6, max_length=20)


class SmsTestResult(BaseModel):
    ok: bool
    message: str
